cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
   }
 
   public void upgradeVersionKeys() {
-    MDSKey startKey = new MDSKey.Builder().add(TYPE_APP_META).build();
-    Map<MDSKey, ApplicationMeta> appMetas = listKV(startKey, ApplicationMeta.class);
-    Map<MDSKey, ApplicationMeta> newAppMetas = new HashMap<>();
-    for (Map.Entry<MDSKey, ApplicationMeta> oldKey : appMetas.entrySet()) {
-      MDSKey newKey = appendDefaultVersion(oldKey.getKey());
-      newAppMetas.put(newKey, oldKey.getValue());
+    upgradeVersionKeys(TYPE_APP_META, ApplicationMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_STARTED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_SUSPENDED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_COMPLETED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_NODE_STATE, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_TOKEN, RunRecordMeta.class);
+  }
+
+  public <T> void upgradeVersionKeys(String recordType, Type typeOfT) {
+    MDSKey startKey = new MDSKey.Builder().add(recordType).build();
+    Map<MDSKey, T> oldMap = listKV(startKey, typeOfT);
+    Map<MDSKey, T> newMap = new HashMap<>();
+    for (Map.Entry<MDSKey, T> oldEntry : oldMap.entrySet()) {
+      MDSKey newKey = appendDefaultVersion(recordType, typeOfT, oldEntry.getKey());
+      newMap.put(newKey, oldEntry.getValue());
     }
 
     //Delete old rows
     deleteAll(startKey);
 
     // Write new rows
-    for (Map.Entry<MDSKey, ApplicationMeta> newMeta : newAppMetas.entrySet()) {
-      write(newMeta.getKey(), newMeta.getValue());
+    for (Map.Entry<MDSKey, T> newEntry : newMap.entrySet()) {
+      write(newEntry.getKey(), newEntry.getValue());
     }
   }
 
-  // Append version only if it doesn't have version at the end
-  private static MDSKey appendDefaultVersion(MDSKey oldKey) {
-    MDSKey.Splitter splitter = oldKey.split();
-    int keyParts = 0;
+  // Append version only if it doesn't have version in the key
+  private static MDSKey appendDefaultVersion(String recordType, Type typeOfT, MDSKey originalKey) {
+    List<String> keyParts = new ArrayList<>();
+    MDSKey.Splitter splitter = originalKey.split();
     try {
-      // max parts = 4 => appMeta.ns.app.version
-      for (int i = 0; i < 4; i++) {
-        splitter.skipString();
-        keyParts++;
+      // try to iterate through all key parts of type String
+      while (true) {
+        keyParts.add(splitter.getString());
```

Comment:
Is it guaranteed that `splitter.getString()` will throw exception? What happens if it does not?